File: SM3/sm3.cpp
#include"sm3.h"
#include<Windows.h>
#include<iostream>
#include<immintrin.h>



LARGE_INTEGER nFreq;
LARGE_INTEGER nBeginTime;
LARGE_INTEGER nEndTime;
double Time = 0;

static unsigned char message_buffer[64] = { 0 };
static unsigned int iv[8] = { 0 };
static unsigned int T[64] = { 0 };

/* set the T*/
int init_T() {
	for (int i = 0; i < 16; i++)
		T[i] = 0x79cc4519;
	for (int i = 16; i < 64; i++)
		T[i] = 0x7a879d8a;
	return 1;
}


void SM3_init()
{
	init_T();
	iv[0] = 0x7380166f;
	iv[1] = 0x4914b2b9;
	iv[2] = 0x172442d7;
	iv[3] = 0xda8a0600;
	iv[4] = 0xa96f30bc;
	iv[5] = 0x163138aa;
	iv[6] = 0xe38dee4d;
	iv[7] = 0xb0fb0e4e;
}



void out_hex()
{
	unsigned int i = 0;
	for (i = 0; i < 8; i++)
	{
		printf("%08x ", iv[i]);
	}
	printf("\n");
}

/*define the bool function*/
unsigned int FF(unsigned int X, unsigned int Y, unsigned int Z, unsigned int j) {
	unsigned int ret = 0;
	if (0 <= j && j < 16)
		ret = X ^ Y ^ Z;
	else if (16 <= j && j < 64)
		ret = (X & Y) | (X & Z) | (Y & Z);
	return ret;
}


unsigned int GG(unsigned int X, unsigned int Y, unsigned int Z,unsigned int j) {
	unsigned int ret = 0;
	if (0 <= j && j < 16)
		ret = X ^ Y ^ Z;
	else if (16 <= j && j < 64)
		ret = (X & Y) | ((~X) & Z);
	return ret;
}

unsigned int Round_shift(unsigned int a, unsigned int k) {
	k = k % 32;
	return ((a << k) & 0xFFFFFFFF) | ((a & 0xFFFFFFFF) >> (32 - k));
}
/*The function is for round_shift
*k is the number of shift
* the xor is using for the extend

*/


#define P_0(X) X^(Round_shift(X,9))^(Round_shift(X,17))
#define P_1(X) X^(Round_shift(X,15))^(Round_shift(X,23))



int IC(unsigned char* arr) {
	unsigned int W[68];
	unsigned int _W[64];
	unsigned int j;
	unsigned int A, B, C, D, E, F, G, H;
	unsigned int SS1, SS2, TT1, TT2;
	for (j = 0; j < 16; j++)
		W[j] = arr[j * 4 + 0] << 24 | arr[j * 4 + 1] << 16 | arr[j * 4 + 2] << 8 | arr[j * 4 + 3];
	for (j = 16; j < 68; j++)
		W[j] = P_1(W[j - 16] ^ W[j - 9] ^ (Round_shift(W[j - 3], 15))) ^ (Round_shift(W[j - 13], 7)) ^ W[j - 6];
	for (j = 0; j < 64; j++)
		_W[j] = W[j] ^ W[j + 4];
	A = iv[0];
	B = iv[1];
	C = iv[2];
	D = iv[3];
	E = iv[4];
	F = iv[5];
	G = iv[6];
	H = iv[7];
	for (int j = 0; j < 64; j++)
	{
		SS1 = Round_shift(((Round_shift(A, 12)) + E + (Round_shift(T[j], j))) & 0xFFFFFFFF,7);
		SS2 = SS1 ^ Round_shift(A, 12);
		TT1 = (FF(A, B, C, j) + D + SS2 + _W[j])&0xFFFFFFFF;
		TT2 = (GG(E, F, G, j) + H + SS1 + W[j])&0xFFFFFFFF;
		D = C;
		C = Round_shift(B, 9);
		B = A;
		A = TT1;
		H = G;
		G = Round_shift(F, 19);
		F = E;
		E = P_0(TT2);

	}
	iv[0] = (A ^ iv[0]);
	iv[1] = (B ^ iv[1]);
	iv[2] = (C ^ iv[2]);
	iv[3] = (D ^ iv[3]);
	iv[4] = (E ^ iv[4]);
	iv[5] = (F ^ iv[5]);
	iv[6] = (G ^ iv[6]);
	iv[7] = (H ^ iv[7]);
	return 1;
}

/*Group the messages and iteratively compress the filled chunks*/
void Block(unsigned char* msg, unsigned int msglen) {
	int i;
	int left = 0;
	unsigned long long total = 0;

	for (i = 0; i < msglen / 64; i++) {
		memcpy(message_buffer, msg + i * 64, 64);
		IC(message_buffer);
	}
	/*First fill the each block*/
	total = msglen * 8;
	/*every char has 8 bit*/
	left = msglen % 64;
	memset(&message_buffer[left], 0, 64 - left);
	/*memset(address,value,sizeof(the lenth of address))
	 The memset function is an initialization function that initializes a contiguous piece of memory to a value.

	 It is initialized in bytes.
	*/
	memcpy(message_buffer, msg + i * 64, left);
	/*now the value of i is the last chunk*/
	message_buffer[left] = 0x80;
	if (left <= 55) {
		for (i = 0; i < 8; i++)
			message_buffer[56 + i] = (total >> ((8 - 1 - i) * 8)) & 0xFF;
      		IC(message_buffer);
	}
	else {
		IC(message_buffer);
		memset(message_buffer, 0, 64);
		for (i = 0; i < 8; i++)
		{
			message_buffer[56 + i] = (total >> (8 - 1 - i) * 8) & 0xFF;

		}
		IC(message_buffer);
	}
}


int SM3(unsigned char* text, int len, unsigned char* iv) {
	SM3_init();
	Block(text, len);
	return 1;
}


int main(int argc, int* argv[]) {
	unsigned char iv[32] = { 0 };
	const char* text = "abc";
	int len;
	len = strlen(text);
	QueryPerformanceFrequency(&nFreq);
	QueryPerformanceCounter(&nBeginTime);
	SM3((unsigned char*)text, len, iv);
	QueryPerformanceCounter(&nEndTime);
	Time = (double)(nEndTime.QuadPart - nBeginTime.QuadPart) / (double)nFreq.QuadPart;
	std::cout << "Total time is  " << Time << std::endl;
	out_hex();
	return 0;
}

File: SM3/SM3.py
# SM3
import math
from typing import ByteString


IV = [0x7380166F, 0x4914B2B9, 0x172442D7, 0xDA8A0600, 0xA96F30BC, 0x163138AA, 0xE38DEE4D, 0xB0FB0E4E]
T = [0x79cc4519, 0x7a879d8a]

def AsToByte(string):
    BString = ''
    for i in string:
        BString += hex(ord(i))[2:]
    return BString


def FF(X, Y, Z, j):
    if j >= 0 and j <= 15:
        return X ^ Y ^ Z
    else:
        return ((X & Y) | (X & Z) | (Y & Z))


def RoundS(X, i):
    i = i % 32
    return ((X << i) & 0xFFFFFFFF) | ((X & 0xFFFFFFFF) >> (32 - i))


def GG(X, Y, Z, j):
    if j >= 0 and j <= 15:
        return X ^ Y ^ Z
    else:
        return ((X & Y) | (~X & Z))


def P0(X):
    return X ^ RoundS(X, 9) ^ RoundS(X, 17)


def P1(X):
    return X ^ RoundS(X, 15) ^ RoundS(X, 23)


def T_(j):
    if j >= 0 and j <= 15:
        return T[0]
    else:
        return T[1]


def Pad(message):
    m = bin(int(message, 16))[2:]
    if len(m) != len(message) * 4:
        m = '0' * (len(message) * 4 - len(m)) + m
    l = len(m)
    Pad_len = '0' * (64 - len(bin(l)[2:])) + bin(l)[2:]
    m = m + '1'
    if len(m) % 512 > 448:
        m = m + '0' * (512 - len(m) % 512 + 448) + Pad_len
    else:
        m = m + '0' * (448 - len(m) % 512) + Pad_len
    m = hex(int(m, 2))[2:]
    return m


def Group(m):
    n = len(m) / 128
    M = []
    for i in range(int(n)):
        M.append(m[0 + 128 * i:128 + 128 * i])
    return M


def Ex_msg(M, n):
    W = []
    _W = []
    for j in range(16):
        W.append(int(M[n][0 + 8 * j:8 + 8 * j], 16))
    for j in range(16, 68):
        W.append(P1(W[j - 16] ^ W[j - 9] ^ RoundS(W[j - 3], 15)) ^ RoundS(W[j - 13], 7) ^ W[j - 6])
    for j in range(64):
        _W.append(W[j] ^ W[j + 4])
    return W, _W


def CF(V, M, i):
    A, B, C, D, E, F, G, H = V[i]
    W, _W = Ex_msg(M, i)
    for j in range(64):
        SS1 = RoundS((RoundS(A, 12) + E + RoundS(T_(j), j % 32)) % (2 ** 32), 7)
        SS2 = SS1 ^ RoundS(A, 12)
        TT1 = (FF(A, B, C, j) + D + SS2 + _W[j]) % (2 ** 32)
        TT2 = (GG(E, F, G, j) + H + SS1 + W[j]) % (2 ** 32)
        D = C
        C = RoundS(B, 9)
        B = A
        A = TT1
        H = G
        G = RoundS(F, 19)
        F = E
        E = P0(TT2)
    a, b, c, d, e, f, g, h = V[i]
    V_ = [a ^ A, b ^ B, c ^ C, d ^ D, e ^ E, f ^ F, g ^ G, h ^ H]
    return V_


def Round_iter(M):
    n = len(M)
    V = []
    V.append(IV)
    for i in range(n):
        V.append(CF(V, M, i))
    return V[n]


def SM3(message):
    m = Pad(message)  
    M = Group(m)  
    Vn = Round_iter(M)  
    res = ''
    for x in Vn:
            res += hex(x)[2:]
    return res



def KDF(z,klen):
    klen = int(klen)
    ct = 0x00000001
    cnt = math.ceil(klen/32)
    Ha = ""
    for i in range(cnt):
        msg = z+hex(ct)[2:].rjust(8,'0')
        #print(msg)
        Ha  = Ha + SM3(msg)
        #print(Ha)
        ct += 1
    return Ha[0:klen*2]


if __name__ == "__main__":
    content = AsToByte('abc')
    res = SM3(content)
    print(res)